src/library/scala/collection/TraversableOnce.scala

```diff
@@ -115,32 +116,39 @@ trait TraversableOnce[+A] extends Any with GenTraversableOnce[A] {
 
   // for internal use
   protected[this] def reversed = {
-    object reverser extends Function1[A, Unit] {
+    //avoid the LazyRef as we don't have an @eager object
+    class reverser extends AbstractFunction1[A, Unit] {
       var elems: List[A] = Nil
       override def apply(v1: A): Unit = elems ::= v1
     }
+    val reverser = new reverser
     self foreach reverser
     reverser.elems
   }
 
   def size: Int = {
     //we can't guard with isEmpty as some implementation have
     // def isEmpty = size == 0
-    object counter extends Function1[A, Unit] {
+
+    //avoid the LazyRef as we don't have an @eager object
+    class counter extends AbstractFunction1[A, Unit] {
       var result = 0
       override def apply(v1: A): Unit = result += 1
     }
+    val counter = new counter
```

Comment:
what do you think about 
```suggestion
    val counter = new AbstractFunction1[A, Unit] {
      var result = 0
      override def apply(v1: A): Unit = result += 1
    }
```

Reply:
The problem with you suggestion is that `result` is not part of the signature if `AbstractFunction1`and is therefore read by reflection code that the compiler generates
```
   LINENUMBER 140 L5
    ALOAD 1
    ASTORE 2
   L0
    ALOAD 2
    INVOKEVIRTUAL scala/runtime/AbstractFunction1.getClass ()Ljava/lang/Class;
    INVOKESTATIC scala/collection/TraversableOnce.reflMethod$Method1 (Ljava/lang/Class;)Ljava/lang/reflect/Method; (itf)
    ALOAD 2
    ICONST_0
    ANEWARRAY java/lang/Object
    INVOKEVIRTUAL java/lang/reflect/Method.invoke (Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;
   L1
    GOTO L6
   L2
   FRAME FULL [scala/collection/TraversableOnce scala/collection/TraversableOnce$$anon$1 scala/collection/TraversableOnce$$anon$1] [java/lang/reflect/InvocationTargetException]
    ASTORE 3
    ALOAD 3
    INVOKEVIRTUAL java/lang/reflect/InvocationTargetException.getCause ()Ljava/lang/Throwable;
    ATHROW
   L6
   FRAME SAME1 java/lang/Object
    CHECKCAST java/lang/Integer
   L7
    INVOKESTATIC scala/runtime/BoxesRunTime.unboxToInt (Ljava/lang/Object;)I
   L8
    IRETURN
```

Reply:
Ah, yes!